crates/mdns/src/record.rs

```diff
@@ -0,0 +1,294 @@
+                for other_rec in other_records.iter() {
+                    cached_records.push(other_rec.clone());
+                }
+
+                cached_records.sort();
+
+                let mut uniq_urls: Vec<String> =
+                    cached_records.iter().map(|rec| rec.url.clone()).collect();
+
+                uniq_urls.dedup();
+
+                let mut final_records: Vec<Record> = Vec::new();
+
+                for url in uniq_urls {
+                    let drain_record_with_url: Vec<Record> = cached_records
+                        .drain_filter(|rec| *rec.url == *url)
+                        .collect();
+                    if drain_record_with_url.len() == 1 {
+                        final_records.push(drain_record_with_url[0].clone());
+                    } else {
+                        if let Some(first_record) = drain_record_with_url.first() {
+                            if first_record.ttl == 0 {
+                                final_records.push(first_record.clone());
+                            }
+                        }
+                        if let Some(last_record) = drain_record_with_url.last() {
+                            if last_record.ttl == 255 {
+                                final_records.push(last_record.clone());
+                            }
+                        }
+                    }
+                }
+
+                *cached_records = final_records.clone();
+            } else {
+                self.0.insert(other_nid.to_string(), other_records.to_vec());
+            }
+        }
+    }
+
+    /// Builds a [`MapRecord`] from a [`DnsMessage`].
+    pub fn from_dns_message(dmesg: &DnsMessage) -> Option<MapRecord> {
+        if !dmesg.answers.is_empty() {
+            let mut records: Vec<Record> = dmesg
+                .answers
+                .iter()
+                .map(|a_sec| Record::new(&a_sec.domain_name, &a_sec.data.target, a_sec.ttl))
+                .collect();
+
+            records.sort();
+            records.dedup_by(|a, b| a.url() == b.url());
+
+            let mut map_record = MapRecord(HashMapRecord::with_capacity(records.len()));
+            for new_record in records.iter() {
+                let fake_map_record =
+                    MapRecord::with_record(&new_record.networkid, &[new_record.clone()]);
+                map_record.update(&fake_map_record);
+            }
+
+            Some(map_record)
+        } else {
+            None
+        }
+    }
+
+    /// Builds a [`DnsMessage`] from a [`MapRecord`].
+    pub fn to_dns_response_message(&self, networkids: &[&str]) -> Option<DnsMessage> {
+        // Let's make sure we have at least one networkid in our keys
+        let mut answers = Vec::new();
+
+        for &netid in networkids.iter() {
+            if let Some(records) = self.0.get(netid) {
+                for rec in records {
+                    answers.push(rec.to_answers_section());
```

Comment:
I was imagining only responding with our own record, but I'm cool with this :+1: If we find our packets are getting too big, we can always re-evaluate.